packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,265 @@
+
+      const compared = new Set<IPreprocessor>();
+
+      if (exclusivePreprocessor.getConditions().length) {
+        // If we find a duplicate preprocessor, we will use the existing one instead.
+        for (const existingPreprocessor of this.envConditionMap.values()) {
+          if (!compared.has(existingPreprocessor) && compare(preprocessor, existingPreprocessor)) {
+            exclusivePreprocessor = existingPreprocessor;
+
+            break;
+          }
+
+          compared.add(existingPreprocessor);
+        }
+
+        preserveFilterIds.add(filterId);
+        this.envConditionMap.set(filterId, exclusivePreprocessor);
+      }
+    }
+
+    for (const [filterId, preprocessor] of newPreprocessors.entries()) {
+      // If we encounter the duplicate filter id, we will assign new preprocessor
+      // handling two conditions.
+      if (this.envConditionMap.has(filterId)) {
+        const existingPreprocessor = this.envConditionMap.get(filterId)!;
+
+        if (compare(preprocessor, existingPreprocessor)) {
+          this.envConditionMap.set(filterId, existingPreprocessor);
+
+          continue;
+        }
+
+        let mergedPreprocessor: IPreprocessor;
+
+        if (existingPreprocessor instanceof NegatedPreprocessor) {
+          mergedPreprocessor = new NegatedPreprocessor({
+            ref: new Preprocessor({
+              conditions: existingPreprocessor.getConditions(),
+              rawLine: existingPreprocessor.ref.rawLine,
+            }),
+            rawLine: existingPreprocessor.rawLine,
+          });
+        } else {
+          mergedPreprocessor = new Preprocessor({
+            conditions: existingPreprocessor.getConditions(),
+            rawLine: preprocessor.rawLine,
+          });
+        }
+
+        for (const condition of preprocessor.getConditions()) {
+          mergedPreprocessor.addCondition(condition);
+        }
+
+        this.envConditionMap.set(filterId, mergedPreprocessor);
+
+        continue;
+      }
+
+      // If we find a duplicate filter id, we will use the existing one instead.
+      const compared = new Set<IPreprocessor>();
+
+      let ref: IPreprocessor = preprocessor;
+
+      for (const existingPreprocessor of this.envConditionMap.values()) {
+        if (!compared.has(existingPreprocessor) && compare(preprocessor, existingPreprocessor)) {
+          ref = existingPreprocessor;
+
+          break;
+        }
+
+        compared.add(existingPreprocessor);
+      }
+
+      this.envConditionMap.set(filterId, ref);
+    }
+
+    return {
+      preserveFilterIds,
+    };
+  }
+
+  public isEnvQualifiedFilter(filter: IFilter) {
+    const preprocessor = this.envConditionMap.get(filter.getId());
+
+    if (!preprocessor) {
+      return true;
+    }
+
+    let result = this.cache.get(preprocessor);
+
+    if (!result) {
+      result = preprocessor.evaluate(this.env.mask);
+
+      this.cache.set(preprocessor, result);
+    }
+
+    return result;
+  }
+
+  public invertMap() {
```

Comment:
If possible, we need a better approach here in both performance wise and complexity wise.